### PR TITLE
Pass Deploy_groups into Deploy script for use.

### PR DESCRIPTION
**Pull request by henders**
To further the integration of deploy groups, this passes down the list of affected deploy_groups in an environment variable $DEPLOY_GROUPS so the script can access it.

screen shot 2015-05-08 at 15 05 18

----------------------

pod1_deploy__succeeded__-_project0

/cc @zendesk/runway @zendesk/samson 

### References
 - Jira link:

### Risks
 - None

### CODE REVIEW

**Comment by grosser** [comment]
don't they need to be lowercase to be used in cap ?

File: app/models/job_execution.rb [hunk 1]
```diff
@@ -101,6 +101,7 @@ def execute!(dir)
 
     FileUtils.mkdir_p(artifact_cache_dir)
     @output.write("Executing deploy\n")
+    @output.write("Deploy URL: #{@job.deploy.url}\n") if @job.deploy
```
Comment (grosser) [inline app/models/job_execution.rb]: Isn't the log only shown on the deploy page anyway ?

File: app/models/job_execution.rb [hunk 2]
```diff
@@ -109,6 +110,7 @@ def execute!(dir)
       "export REVISION=#{@reference.shellescape}",
       "export TAG=#{(@job.tag || @job.commit).to_s.shellescape}",
       "export CACHE_DIR=#{artifact_cache_dir}",
+      "export DEPLOY_GROUPS='#{@stage.deploy_groups.map(&:name).join(' ') if @stage}'",
```
Comment (grosser) [inline app/models/job_execution.rb]: setting an empty set would be a little bit of clutter for non-deploy-grouped deploys
->

```Ruby
xxx = []
if @stage && deploy_groups = @stage.deploy_groups.map(&:name).join(' ').presence
  xxx << "export DEPLOY_GROUPS='#{deploy_groups}'"
end
```

**Comment by grosser** [comment]
:+1: except for the also exporting empty

**Comment by henders** [comment]
@grosser , yes, but I figure the script can take care of lowercasing them. Didn't want to change the case if someone else besides Zendesk needs it to exact match the DeployGroup name.

So they would use:
```
PODS = `echo $DEPLOY_GROUPS | | awk '{print tolower($0)}'`
cap $PODS deploy
```

File: app/models/job_execution.rb (hunk 1 above)
Comment (henders) [inline app/models/job_execution.rb]: Yea, @pswadi-zendesk requested this for auditing purposes in the runway flow!

**Comment by grosser** [comment]
hmm technically more correct, but a bit messy ... we can just change our deploy groups to be lowercase I guess.

File: app/models/job_execution.rb (hunk 2 above)
Comment (sandlerr) [inline app/models/job_execution.rb]: probably also want to shell escape these. Not sure what would happen if they had a space or single quote

Comment (henders) [inline app/models/job_execution.rb]: Good point! I'll do that.

**Comment by jonmoter** [comment]
Hey @henders, can you remind me of the use case for this? I know we spoke about it at some point, but I've forgotten how this will be used.

**Comment by henders** [comment]
@jonmoter 
It's integrating DeployGroup functionality further into Samson. At the moment it's just visual fluff, but this change allows the stage to just do 'cap $DEPLOY_GROUPS deploy' instead of enumerating the deploy groups manually when they've already ticked the appropriate deploy group boxes.

**Comment by henders** [comment]
@zendesk/runway @grosser 
Looking for another +1 on this since I refactored some code which still looks a bit off.

File: test/models/job_execution_test.rb [hunk 3]
```diff
@@ -125,6 +127,15 @@
     lines.must_include "TAG=v1"
   end
 
+  it 'exports deploy_group information if deploy groups present' do
+    job.update(command: 'env')
+    job.deploy = deploy
+    execute_job
+    lines = job.output.split "\n"
+    lines.must_include "# Deploy URL: #{deploy.url}"
+    lines.must_include "DEPLOY_GROUPS=Pod1 Pod2"
```
Comment (jonmoter) [inline test/models/job_execution_test.rb]: Shouldn't there be single-quotes surrounding "Pod1 Pod2"?

File: app/models/job_execution.rb [hunk 4]
```diff
@@ -101,28 +101,18 @@ def execute!(dir)
...
-      "cd #{dir}",
-      *@job.commands
-    ]
+    @output.write("# Deploy URL: #{@job.deploy.url}\n") if @job.deploy
 
     payload = {
       stage: (stage.try(:name) || "none"),
       project: @job.project.name,
-      command: commands.join("\n")
+      command: commands(dir).join("\n")
```
Comment (grosser) [inline app/models/job_execution.rb]: hmm calling commands twice is a bit strange ... maybe store it in some local variable ?

Comment (grosser) [inline app/models/job_execution.rb]: yeah ... it's cached but has an argument ... that smells like trouble :)
-> no cache but store as local variable

File: app/models/job_execution.rb [hunk 5]
```diff
@@ -142,6 +132,25 @@ def setup!(dir)
...
+        "export DEPLOYER=#{@job.user.email.shellescape}",
+        "export DEPLOYER_EMAIL=#{@job.user.email.shellescape}",
+        "export DEPLOYER_NAME=#{@job.user.name.shellescape}",
+        "export REVISION=#{@reference.shellescape}",
+        "export TAG=#{(@job.tag || @job.commit).to_s.shellescape}",
+        "export CACHE_DIR=#{artifact_cache_dir}",
+        "cd #{dir}",
+        *@job.commands
+      ]
+      if @stage && @stage.deploy_groups.count > 0
```
Comment (grosser) [inline app/models/job_execution.rb]: saving an extra query and map! for a bit of performance win:
`if @stage && group_names = @stage.deploy_groups.pluck(:name).map!(&:shellescape).join(" ")`

File: app/models/job_execution.rb [hunk 6]
```diff
@@ -142,6 +132,25 @@ def setup!(dir)
...
+        commands.insert(0, "export DEPLOY_GROUPS='#{@stage.deploy_groups.map(&:name).map(&:shellescape).join(' ')}'")
```
Comment (grosser) [inline app/models/job_execution.rb]: FYI insert 0 == unshift

File: test/models/job_execution_test.rb [hunk 7]
```diff
@@ -125,6 +127,15 @@
...
+  end
+
```
Comment (grosser) [inline test/models/job_execution_test.rb]: a test shellescape ?

Comment (grosser) [inline test/models/job_execution_test.rb]: test for not having deploy groups -> no DEPLOY_GROUPS set

**Comment by grosser** [comment]
we'll have to change our names deploy group names to lowercase to be able to use them with cap then or lowercase them in cap ?

**Comment by gabetax** [comment]
We could leave it to the command to do any translation necessary: http://stackoverflow.com/a/2264537/119479 - janky, but flexible.

Alternatively, we could introduce this concept within the `HostGroup` model as a value that can be explicitly defined.

**Comment by grosser** [comment]
meh just renaming them seems easier :)

File: test/models/job_execution_test.rb (hunk 3 above)
Comment (henders) [inline test/models/job_execution_test.rb]: This is matching against the output of 'env' which won't have the quotes in them.

**Comment by henders** [comment]
@grosser - addressed comments!

File: app/models/job_execution.rb (hunk 6 above)
Comment (henders) [inline app/models/job_execution.rb]: Ah, thanks! :)

File: app/models/job_execution.rb (hunk 4 above)
Comment (henders) [inline app/models/job_execution.rb]: Yea, good call... wasn't liking this smell.

File: app/models/job_execution.rb [hunk 8]
```diff
@@ -142,6 +133,23 @@ def setup!(dir)
...
+      "export DEPLOYER=#{@job.user.email.shellescape}",
+      "export DEPLOYER_EMAIL=#{@job.user.email.shellescape}",
+      "export DEPLOYER_NAME=#{@job.user.name.shellescape}",
+      "export REVISION=#{@reference.shellescape}",
+      "export TAG=#{(@job.tag || @job.commit).to_s.shellescape}",
+      "export CACHE_DIR=#{artifact_cache_dir}",
+      "cd #{dir}",
+      *@job.commands
+    ]
+    if @stage && group_names = @stage.deploy_groups.pluck(:name).map!(&:shellescape).join(" ")
```
Comment (grosser) [inline app/models/job_execution.rb]: `.presence` or it will assign an empty string when empty

File: test/models/job_execution_test.rb [hunk 9]
```diff
@@ -123,6 +126,16 @@
...
+    job.output.wont_include "DEPLOY_GROUPS"
+  end
+
+  it 'exports deploy_group information if deploy groups present' do
+    job.update(command: 'env | sort')
+    job.deploy = deploy
+    execute_job
+    lines = job.output.split "\n"
+    lines.must_include "# Deploy URL: #{deploy.url}"
+    lines.must_include 'DEPLOY_GROUPS=\;\|sudo\ make-sandwich\ / Pod1 Pod2'
```
Comment (grosser) [inline test/models/job_execution_test.rb]: a `it "does not add deploy group if deploy groups are missing"` would be nice :)

**Comment by grosser** [comment]
:+1:

**Pull request merged**